csrc/host_ir/jit.cpp

```diff
@@ -166,29 +178,388 @@ void printLlvmIr(llvm::Function* func, std::string_view msg) {
+      // currently we just pad inner value, so dividend is outer value
+      // so inner_extent = (out_extent + outer_extent - 1) / outer_extent, which is a ceilDiv
+      llvm::Value* outer_extent_value = getOrCreateValueForExtent(merge->outer(), val_to_value, builder);
+      llvm::Value* minus_one = builder.CreateSub(outer_extent_value, builder.getInt64(1));
+      llvm::Value* plus_value = builder.CreateAdd(out_extent_value, minus_one);
+      llvm::Value* inner_extent_value = builder.CreateUDiv(plus_value, outer_extent_value);
+        
+      level_order_domains.insert(out_i, merge->outer(), outer_extent_value);
+      level_order_domains.insert(out_i, merge->inner(), inner_extent_value);
+    } else {
+      NVF_THROW("LLVM Lowering Error: Unsupported expression type: ", transform->getOpString());
+    }
+  }
+
+  // This should contains same iter domains as logical domain, but in different order
+  std::vector<IterDomain*> propogated_allocation_domains;
+  for(auto it : level_order_domains) {
+    propogated_allocation_domains.push_back(it.first);
+  }
+
+  auto permutation = ir_utils::computePermutation(
+    logical_domain, propogated_allocation_domains);
+  NVF_ERROR(permutation.has_value(), "LLVM Lowering Error: Failed to compute permutation");
+  
+  // Map last level propagated allocation domains to logical domain
+  // we should be able to get the permutation between them
+  llvm::Value* alter_stride_value = builder.getInt64(1);
+  std::vector<llvm::Value*> allocation_sizes_values;
+  std::vector<llvm::Value*> allocation_strides_values;
+  for(auto it : propogated_allocation_domains | std::views::reverse) {
+    if(it->isDeviceDim()) {
+      // Dummy value, we will filter out this dimension in the end
+      allocation_sizes_values.push_back(builder.getInt64(-1));
+      allocation_strides_values.push_back(builder.getInt64(-1));
+    }
+    else if(it->isBroadcast()) {
+      allocation_sizes_values.push_back(getOrCreateValueForExtent(it, val_to_value, builder));
+      allocation_strides_values.push_back(builder.getInt64(0));
+    }
+    else{
+      auto [extent_value, out_i] = level_order_domains.erase(it);
+      level_order_domains.insert(out_i, it, extent_value);
+      allocation_sizes_values.push_back(extent_value);
+      allocation_strides_values.push_back(alter_stride_value);
+      alter_stride_value = builder.CreateMul(alter_stride_value, extent_value);
+    }
+  }
+  std::reverse(allocation_sizes_values.begin(), allocation_sizes_values.end());
+  std::reverse(allocation_strides_values.begin(), allocation_strides_values.end());
+
+  sizes.resize(allocation_sizes_values.size());
+  strides.resize(allocation_strides_values.size());
+  
+  // Apply permutation correctly by mapping from allocation domain order to logical domain order
+  for(size_t i = 0; i < allocation_sizes_values.size(); ++i) {
+    size_t logical_idx = permutation.value()[i];
+    sizes[logical_idx] = allocation_sizes_values[i];
```

Comment:
This one felt slightly wrong. Allocation size and size aren't the same thing.

i.e. if we pad something because of split, they shouldn't be reflected on the sizes of the logical tensor.

Reply:
You are right. To properly fix this (in a follow-up PR), I think we need to up-propagate **two** types of sizes to the logical domain, (local) logical sizes and (local) allocation sizes. The former are same as logical dimension sizes except that certain dimension sizes are divided by the number of devices along a certain mesh axis. The latter are used for computing strides.

Reply:
There's also a limitation on up-propagating via merges with DIDx presence. I don't know how to fix that. Consider the following two examples:

```
 2   4
  \ /
   8
  / \
 4   2
DIDx
```
where the local sizes should be [1, 2]

```
 2   4
  \ /
   8
  / \
 2   4
   DIDx
```
where the local sizes should be [2, 1]

The current algorithm can't even distinguish the two cases when it's at `merge(2, 4)`. https://github.com/NVIDIA/Fuser/blob/ef0696335b0ee893a84d95a7180bf00b6e41659b/csrc/runtime/allocations.cpp#L476 is wrong as well but in a slightly different way.